firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1035 @@
+over, to result in::
+
+   [[0, 1, 3, 4],
+    [4, 5, 6, 7],
+    [7, 8, 9, 10]]
+
+Now, when we iterate over cells, we ensure that we access the correct
+dofs.  The same trick needs to be applied to facet maps too.
+
+Computation of boundary nodes
+-----------------------------
+
+For the top and bottom boundary nodes, we walk over the cells at,
+respectively, the top and bottom of the column and pull out those
+nodes whose entity height matches the appropriate cell height.  As an
+example::
+
+                      8--------10
+                      |        |
+                      |        |
+                      |        |
+                      |        |
+    2--------5--------7--------9
+    |        |        |
+    |        |        |
+    |        |        |
+    |        |        |
+    1--------4--------6
+    |        |
+    |        |
+    |        |
+    |        |
+    0--------3
+
+The bottom boundary nodes are::
+
+   [0, 3, 4, 6, 7, 9]
+
+whereas the top are::
+
+   [2, 5, 7, 8, 10]
+
+For these strange "interior" facets, we first walk over the cells,
+picking up the dofs in the closure of the base (ceiling) of the cell,
+then we walk over facets, picking up all the dofs in the closure of
+facets that are exposed (there may be more than one of these in the
+cell column).  We don't have to worry about any lower-dimensional
+entities, because if a co-dim 2 or greater entity is exposed in a
+column, then the co-dim 1 entity in its star is also exposed.
+
+For the side boundary nodes, we can make a simplification: we know
+that the facet heights are always the same as the cell column heights
+(because there is only one cell in the support of the facet).  Hence
+we just walk over the boundary facets of the base mesh, extract out
+the nodes on that facet on the bottom cell and walk up the column.
+This is guaranteed to pick up all the nodes in the closure of the
+facet column.
+"""
+
+from __future__ import absolute_import, print_function, division
+
+from six import iteritems
+from finat.finiteelementbase import entity_support_dofs
+
+cimport mpi4py.MPI as MPI
+from mpi4py.libmpi cimport MPI_Op_create, MPI_Op_free, MPI_User_function
+from mpi4py import MPI
+from firedrake.petsc import PETSc
+import numpy
+cimport numpy
+import cython
+cimport petsc4py.PETSc as PETSc
+from pyop2.datatypes import IntType
+from pyop2 import op2
+from tsfc.fiatinterface import as_fiat_cell
+
+import firedrake.extrusion_utils as eutils
+numpy.import_array()
+
+include "dmplexinc.pxi"
+
+
+cdef inline void extents_reduce(void *in_, void *out, int *count, MPI.MPI_Datatype *datatype) nogil:
+    cdef:
+        PetscInt *xin = <PetscInt *>in_
+        PetscInt *xout = <PetscInt *>out
+
+    if xin[0] < xout[0]:
+        xout[0] = xin[0]
+    if xin[1] > xout[1]:
+        xout[1] = xin[1]
+    if xin[2] > xout[2]:
+        xout[2] = xin[2]
+    if xin[3] < xout[3]:
+        xout[3] = xin[3]
+
+
+@cython.wraparound(False)
+def layer_extents(mesh):
```

Comment:
Receive `plex`, `cell_numbering`, `cell_extents`, and `comm` instead. That way you don't need to construct the `cell_set` to call this.